<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>DNA Cauldron</title>
    
      <link rel="stylesheet" href="_static/pygments.css">
      <link rel="stylesheet" href="_static/theme.css">
      <link rel="stylesheet" href="_static/sphinx_press_theme.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>

      
      <script src="_static/theme-vendors.js"></script>
      <script src="_static/theme.js" defer></script>
    
  <link rel="index" title="Index" href="genindex.html" />
  <link rel="search" title="Search" href="search.html" />
  <link rel="next" title="Useful classes for scripts" href="ref/user_classes.html" /> 
  </head>

  <body>
    <div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="#" class="home-link">
    
      <img class="logo" src="_static/html_logo.png" alt="doit logo"/>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">

  
    <div class="nav-item">
      <a href="#more-biology-software"
         class="nav-link ">
         Reference
      </a>
    </div>
  
    <div class="nav-item">
      <a href="#more-biology-software"
         class="nav-link ">
         Examples
      </a>
    </div>
  



  
    <div class="nav-item">
      <a href="https://github.com/Edinburgh-Genome-Foundry/DnaCauldron"
        class="nav-link external">
          Github <outboundlink/>
      </a>
    </div>
  

    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            

  
    <div class="nav-item">
      <a href="#more-biology-software"
         class="nav-link ">
         Reference
      </a>
    </div>
  
    <div class="nav-item">
      <a href="#more-biology-software"
         class="nav-link ">
         Examples
      </a>
    </div>
  



  
    <div class="nav-item">
      <a href="https://github.com/Edinburgh-Genome-Foundry/DnaCauldron"
        class="nav-link external">
          Github <outboundlink/>
      </a>
    </div>
  

            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="#more-biology-software">Reference</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="ref/user_classes.html" class="reference internal ">Useful classes for scripts</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="ref/utils.html" class="reference internal ">Useful methods built on top of DNA Cauldron</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="ref/biotools.html" class="reference internal ">Biotools module</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="ref/core_classes.html" class="reference internal ">Core classes</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="ref/spreadsheet_reference.html" class="reference internal ">Spreadsheet Reference</a>

            
          </li>

        
      </ul>
    </div>
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="#more-biology-software">Examples</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="examples.html" class="reference internal ">Examples</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="#">Docs</a> &raquo;</li>
    
    <li>DNA Cauldron</li>
  </ul>
  

  <ul class="page-nav">
  <li class="next">
    <a href="ref/user_classes.html"
       title="next chapter">Useful classes for scripts →</a>
  </li>
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <p align="center">
<img alt="DNA Cauldron Logo" title="DNA Cauldron Logo" src="https://raw.githubusercontent.com/Edinburgh-Genome-Foundry/DnaCauldron/master/docs/_static/images/title.png" width="500">
<br /><br />
</p><a class="reference external image-reference" href="https://app.travis-ci.com/Edinburgh-Genome-Foundry/DnaCauldron"><img alt="Travis CI build status" src="https://app.travis-ci.com/Edinburgh-Genome-Foundry/DnaCauldron.svg?branch=master" /></a>
<a class="reference external image-reference" href="https://coveralls.io/github/Edinburgh-Genome-Foundry/DnaCauldron?branch=master"><img alt="https://coveralls.io/repos/github/Edinburgh-Genome-Foundry/DnaCauldron/badge.svg?branch=master" src="https://coveralls.io/repos/github/Edinburgh-Genome-Foundry/DnaCauldron/badge.svg?branch=master" /></a>
<div class="section" id="dna-cauldron">
<h1>DNA Cauldron<a class="headerlink" href="#dna-cauldron" title="Permalink to this headline">¶</a></h1>
<p>DNA Cauldron (full documentation <a class="reference external" href="https://edinburgh-genome-foundry.github.io/DnaCauldron/">here</a>)
is a generic cloning simulation framework to predict
final construct sequences and detect assembly flaws. It aims in particular at
automating the simulation and verification of large (and possibly multi-step)
DNA assembly batches, with the idea that, if your assembly plan works with
Cauldron, you’ll get the same results at the bench.</p>
<ul class="simple">
<li><p>Great support for Golden Gate assembly (incl. MoClo, EMMA, Phytobrick, etc.), the simulation is as simple as
<a class="reference external" href="http://cuba.genomefoundry.org/simulate_gg_assemblies">dropping sequence files in a web app</a>.</p></li>
<li><p>Good support for Gibson, Biobrick, BASIC, LCR Assembly.</p></li>
<li><p>Provide genetic parts in any order, with or without annotations, in reverse or direct
sense, linear or circular, single or combinatorial assembly… Cauldron will get it!</p></li>
<li><p>Design flaw detection (missing parts, unwanted restriction sites, wrong overhang designs, etc.).</p></li>
<li><p>If your assembly needs connector parts, Cauldron can select them for you!</p></li>
<li><p>Import batch assembly plans from spreadsheets, including hierarchical (=multi-step) plans.</p></li>
<li><p>Comprehensive reports with constructs sequences, fragment sequences, summary spreadsheets.</p></li>
<li><p>Export multi-file reports in folders, zip files, or in-memory zip files (for use on servers).</p></li>
</ul>
<p align="center">
<img alt="DNA Cauldron Logo" title="DNA Cauldron Logo" src="https://github.com/Edinburgh-Genome-Foundry/DnaCauldron/raw/master/docs/_static/images/reports_elements.png" width="800">
<br /><br />
</p><div class="section" id="usage-tutorial">
<h2>Usage tutorial<a class="headerlink" href="#usage-tutorial" title="Permalink to this headline">¶</a></h2>
<div class="section" id="providing-part-sequences">
<h3>Providing part sequences<a class="headerlink" href="#providing-part-sequences" title="Permalink to this headline">¶</a></h3>
<p>To simulate an assembly, you need first to provide part sequences. In DNA Cauldron, sequences
are managed by a <code class="docutils literal notranslate"><span class="pre">SequenceRepository</span></code>, created as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dnacauldron</span> <span class="k">as</span> <span class="nn">dc</span>

<span class="c1"># Create a repository from BioPython records</span>
<span class="n">repository</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">SequenceRepository</span><span class="p">(</span><span class="n">collections</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;parts&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;part_A&quot;</span><span class="p">:</span> <span class="n">record_A</span><span class="p">,</span> <span class="s2">&quot;part_B&quot;</span><span class="p">:</span> <span class="n">record_B</span><span class="p">}})</span>

<span class="c1"># Or use &quot;import_sequences&quot; to import files from folders, zip files, etc.</span>
<span class="n">repository</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">SequenceRepository</span><span class="p">()</span>
<span class="n">repository</span><span class="o">.</span><span class="n">import_sequences</span><span class="p">(</span><span class="n">folder</span><span class="o">=</span><span class="s2">&quot;my_sequences/&quot;</span><span class="p">,</span> <span class="n">use_file_names_as_ids</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Important:</strong> ensure that the part’s topology (as set in Biopython records at
<code class="docutils literal notranslate"><span class="pre">record.annotations['topology']</span></code>) is accurate. This can be done at import
time by setting <code class="docutils literal notranslate"><span class="pre">topology='linear'</span></code>, <code class="docutils literal notranslate"><span class="pre">topology='circular'</span></code>, or for instance
<code class="docutils literal notranslate"><span class="pre">topology='default_to_linear'</span></code> to use the topology specified by each Genbank and
default to linear if none is specified.</p>
</div>
<div class="section" id="parts-assembly">
<h3>Parts assembly<a class="headerlink" href="#parts-assembly" title="Permalink to this headline">¶</a></h3>
<p>An assembly simply defines which parts of the sequence repository should
be assembled together. Here we use a “Type-2s Restriction” class of assembly,
which will work for all Golden-Gate-style assemblies:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">assembly</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">Type2sRestrictionAssembly</span><span class="p">(</span><span class="n">parts</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;part_A&quot;</span><span class="p">,</span> <span class="s2">&quot;part_B&quot;</span><span class="p">,</span> <span class="s2">&quot;receptor&quot;</span><span class="p">])</span>
<span class="n">simulation</span> <span class="o">=</span> <span class="n">assembly</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">sequence_repository</span><span class="o">=</span><span class="n">repository</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that we could have provided the enzyme in <code class="docutils literal notranslate"><span class="pre">Type2sRestrictionAssembly</span></code> with
<code class="docutils literal notranslate"><span class="pre">enzyme='BsmBI'</span></code>, but it will be auto-selected by Cauldron.</p>
<p>If you want to simulate other restriction-based assembly reactions such as IGEM Biobricks
or Gibson Assembly instead of Type2S restriction, use the corresponding built-in Assembly subclass:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">assembly</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">BioBrickStandardAssembly</span><span class="p">(</span><span class="n">parts</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;part_A&#39;</span><span class="p">,</span> <span class="s1">&#39;part_B&#39;</span><span class="p">])</span>
<span class="n">assembly</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">GibsonAssembly</span><span class="p">(</span><span class="n">parts</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;part_A&#39;</span><span class="p">,</span> <span class="s1">&#39;part_B&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>Now you can explore the results of the simulation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Print the ID and length of the generated construct(s)</span>
<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">simulation</span><span class="o">.</span><span class="n">construct_records</span><span class="p">:</span>
    <span class="nb">print</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">))</span>

<span class="c1"># Get a list of dictionnaries with data on each construct</span>
<span class="n">constructs_data</span> <span class="o">=</span> <span class="n">simulation</span><span class="o">.</span><span class="n">compute_all_construct_data_dicts</span><span class="p">()</span>

<span class="c1"># Write a full report with sequences and figures in a zip.</span>
<span class="n">simulation</span><span class="o">.</span><span class="n">write_report</span><span class="p">(</span><span class="s2">&quot;report.zip&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>DNA Cauldron aims at generating reports as useful as possible to help you
pinpoint any problem when you don’t get the expected number of assemblies.</p>
</div>
<div class="section" id="assembly-plans">
<h3>Assembly Plans<a class="headerlink" href="#assembly-plans" title="Permalink to this headline">¶</a></h3>
<p>An assembly plan is simply defined by a list of assemblies:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define an assembly plan as a list of Assembly objects</span>
<span class="n">assembly_plan</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">AssemblyPlan</span><span class="p">(</span><span class="n">assemblies</span><span class="o">=</span><span class="p">[</span><span class="n">assembly_1</span><span class="p">,</span> <span class="o">...</span><span class="p">])</span>

<span class="c1"># Or import an assembly plan from spreadsheets:</span>
<span class="n">assembly_plan</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">AssemblyPlan</span><span class="o">.</span><span class="n">from_spreadsheet</span><span class="p">(</span>
    <span class="n">spreadsheet</span><span class="o">=</span><span class="s2">&quot;batch_1.csv&quot;</span><span class="p">,</span> <span class="c1"># could also be an xls(x) file</span>
    <span class="n">assembly_class</span><span class="o">=</span><span class="n">dc</span><span class="o">.</span><span class="n">Type2sRestrictionAssembly</span>
<span class="p">)</span>
</pre></div>
</div>
<p>See these different examples for a spreadsheet defining assemblies.</p>
<p>Assembly plans can be hierarchical (i.e. have an assembly’s construct serve as a
part in another assembly). DNA Cauldron will automatically figure out the dependencies
between assemblies and sort the order in which they should be simulated.</p>
<p>The simulation and reporting on an assembly plan is very similar to that of a single assembly:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plan_simulation</span> <span class="o">=</span> <span class="n">assembly_plan</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">sequence_respository</span><span class="p">)</span>

<span class="c1"># Get a list of dictionnaries with data on each construct</span>
<span class="n">plan_simulation</span><span class="o">.</span><span class="n">compute_all_construct_data_dicts</span><span class="p">()</span>

<span class="c1"># Write a detailed report on each assembly and on the plan as a whole</span>
<span class="n">plan_simulation</span><span class="o">.</span><span class="n">write_report</span><span class="p">(</span><span class="s2">&quot;my_assembly_simulation.zip&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>You can install DNA Cauldron through PIP:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>pip install dnacauldron
</pre></div>
</div>
<p>The full installation using <code class="docutils literal notranslate"><span class="pre">dnacauldron[reports]</span></code> is required for report generation.
Alternatively, you can unzip the sources in a folder and type:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>python setup.py install
</pre></div>
</div>
</div>
<div class="section" id="how-it-works">
<h2>How it works<a class="headerlink" href="#how-it-works" title="Permalink to this headline">¶</a></h2>
<p>DNA Cauldron predicts circular constructs by finding circular paths in part
homology graphs, an idea first described in
<a class="reference external" href="http://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-015-0544-x">Pereira et al. (2015)</a>,
and used in the Python library <a class="reference external" href="https://github.com/BjornFJohansson/pydna">PyDNA</a>.</p>
</div>
<div class="section" id="license-mit">
<h2>License = MIT<a class="headerlink" href="#license-mit" title="Permalink to this headline">¶</a></h2>
<p>DNA Cauldron is an open-source software originally written at the <a class="reference external" href="http://www.genomefoundry.io">Edinburgh Genome Foundry</a> by <a class="reference external" href="https://github.com/Zulko">Zulko</a>
and <a class="reference external" href="https://github.com/Edinburgh-Genome-Foundry/DnaCauldron">released on Github</a> under the MIT license (Copyright 2020 Edinburgh Genome Foundry).
Everyone is welcome to contribute!</p>
</div>
<div class="section" id="more-biology-software">
<h2>More biology software<a class="headerlink" href="#more-biology-software" title="Permalink to this headline">¶</a></h2>
<a class="reference external image-reference" href="https://edinburgh-genome-foundry.github.io/"><img alt="https://raw.githubusercontent.com/Edinburgh-Genome-Foundry/Edinburgh-Genome-Foundry.github.io/master/static/imgs/logos/egf-codon-horizontal.png" src="https://raw.githubusercontent.com/Edinburgh-Genome-Foundry/Edinburgh-Genome-Foundry.github.io/master/static/imgs/logos/egf-codon-horizontal.png" /></a>
<p>DNA Cauldron is part of the <a class="reference external" href="https://edinburgh-genome-foundry.github.io/">EGF Codons</a> synthetic biology software suite for DNA design, manufacturing and validation.</p>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<a href="https://twitter.com/share" class="twitter-share-button"
data-text="DNA Cauldron - A Python module for printing with living matter" data-size="large" data-hashtags="Bioprinting">Tweet
</a>
<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';
if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';
fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');
</script>
<iframe src="http://ghbtns.com/github-btn.html?user=Edinburgh-Genome-Foundry&repo=DNA Cauldron&type=watch&count=true&size=large"
allowtransparency="true" frameborder="0" scrolling="0" width="152px" height="30px" margin-bottom="30px"></iframe></div>
</div>


          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
  <li class="next">
    <a href="ref/user_classes.html"
       title="next chapter">Useful classes for scripts →</a>
  </li>
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2020, Edinburgh Genome Foundry.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.4 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
    </div>
    
    
  </body>
</html>